p1/main.cpp

```diff
@@ -0,0 +1,35 @@
+class Foo {
+    private:
+        std::mutex mtx;
+        std::condition_variable cv;
+        int count = 0;
+        bool isPrinting = false;
+        
+        void printNth(std::function<void()> printFunction, int expectedCount) {
+            {
+                std::unique_lock<std::mutex> lock(mtx);
+                cv.wait(lock, [this, expectedCount]{ return (count == expectedCount) && !isPrinting; });
```

Comment:
Let's say we change predicate from  (count == expectedCount) && !isPrinting

to just (count == expectedCount)

Can you explain to everyone why the latter would be incorrect: specifically give an example of execution of a program with specific (chosen by you) orderings of threads which would render incorrect behavior with that change.

Reply:
Without isPrinting, nothing stops a different thread from running after the current thread releases the lock for the first time but before the printFunction is called. This way the calls to print methods might get recorded

Reply:
This is a bit too vague. Since disproving something is as easy as providing a single counterexample - can you show an execution where things break if we don't have && !isPrinting?